Add support for LT8491

jemfgeronimo opened the pull request:
## PR Description

- This adds driver support, and device tree bindings for LT8491 Battery Charge Controller
- Datasheet: [LT8491](https://www.analog.com/media/en/technical-documentation/data-sheets/LT8491.pdf)
- Tested on RPI4 with [DC2703A](https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/dc2703a-a.html#eb-overview)

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [ ] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

jemfgeronimo committed:
dt-bindings: power: supply: add adi,lt8491.yaml

Add documentation for devicetree bindings for LT8491

Signed-off-by: John Erasmus Mari Geronimo <johnerasmusmari.geronimo@analog.com>

jemfgeronimo committed:
power: supply: add driver for LT8491

LT8491 High Voltage Buck-Boost Battery Charge Controller with I2C

Signed-off-by: John Erasmus Mari Geronimo <johnerasmusmari.geronimo@analog.com>

jemfgeronimo committed:
Kconfig.adi: imply CHARGER_LT8491

Add entry for the LT8491 driver.

Signed-off-by: John Erasmus Mari Geronimo <johnerasmusmari.geronimo@analog.com>

nunojsa commented on Documentation/devicetree/bindings/power/supply/adi,lt8491.yaml:
```diff
@@ -0,0 +1,89 @@
+    description: Resistance value of RFBIN2 in ohms.
+
+  adi,rfbin1-ohms:
+    description: Resistance value of RFBIN1 in ohms.
```
Are there any constrains for the resistor values (or default values)?

jemfgeronimo replied:
None.

jemfgeronimo replied:
These are the actual resistor values used in the circuit/PCB.

nunojsa replied:
And do you need all of them for the driver to work? Very odd that you can just arbitrary values in these things. At the very least, 0 should not be possible...

nunojsa reviewed (commented):
Here it goes my first round

jemfgeronimo commented on Documentation/devicetree/bindings/power/supply/adi,lt8491.yaml:
```diff
@@ -0,0 +1,89 @@
+  - adi,rfbout2-ohms
+  - adi,rdaci-ohms
+  - adi,rfbin2-ohms
+  - adi,rfbin1-ohms
```
Yes. These are needed for the measurements of telemetry data such as temperature, voltage, current, and power.

nunojsa replied:
Ok, it seems they are...

jemfgeronimo commented:
V2
- dt-bindings: fixed some property descriptions
- removed `i2c_set_clientdata` and `i2c_get_clientdata`
- changed `sprintf` to `sysfs_emit`
- removed `lt8491_update_telemetry()` helper
- fixed serial number reading
- fixed signedness of temperature values
- added mutex

nunojsa commented on drivers/power/supply/lt8491_charger.c:
```diff
@@ -0,0 +1,406 @@
+
+	ret = sprintf(info->serial_number, "%04x%04x%04x", serial_number[0], serial_number[1], serial_number[2]);
+	if (ret < 0)
+		return ret;
```
Is this more that 80cols?

nunojsa commented on drivers/power/supply/lt8491_charger.c:
```diff
@@ -0,0 +1,406 @@
+
+		return 0;
+	case POWER_SUPPLY_PROP_VOLTAGE_NOW:
+		mutex_lock(&info->lock);
```
use the cleanup.h stuff. Will make things easier :). You just need `guard(mutex)(&st->lock)` and don't care about unlocking. In here, you might need scoped_guard().

nunojsa commented on drivers/power/supply/lt8491_charger.c:
```diff
@@ -0,0 +1,406 @@
+		if (!FIELD_GET(LT8491_TELEM_ACTIVE_MASK, ret)) {
+			ret = i2c_smbus_write_byte_data(info->client,
+						LT8491_CTRL_UPDATE_TELEM_REG,
+						LT8491_UPDATE_TELEM_CMD);
```
nit: alignment preferred on open parenthesis

nunojsa commented on drivers/power/supply/lt8491_charger.c:
```diff
@@ -0,0 +1,406 @@
+	info->client = client;
+	psy_cfg.drv_data = info;
+
+	mutex_init(&info->lock);
```
upstream use `devm_mutex_init()`

nunojsa reviewed (commented):
Alright, this looks fairly ok to me. Still have some doubts about the resistor values but anyways, send this upstream :)

nunojsa commented:
Hi, I'm just checking in on the status of this pull request. Is there anything I can do to help move it forward?
